core/src/main/java/ai/timefold/solver/core/impl/domain/variable/declarative/DefaultShadowVariableSessionFactory.java

```diff
@@ -0,0 +1,185 @@
+                        });
+                    }
+                }
+            }
+
+            for (var alias : declarativeShadowVariableToAliasMap.getOrDefault(fromVariableId, Collections.emptySet())) {
+                var toVariableId = alias.targetVariableId();
+                var sourceVariableId = new VariableId(alias.entityClass(), alias.variableName());
+
+                if (!alias.isDeclarative() && alias.affectGraphEdges()) {
+                    // Exploit the same fact as above
+                    variableReferenceGraph.addBeforeProcessor(sourceVariableId, (graph, toEntity) -> {
+                        alias.targetEntityFunctionStartingFromVariableEntity().accept(toEntity, fromEntity -> {
+                            graph.removeEdge(
+                                    (EntityVariableOrFactReference<Solution_>) graph.lookup(fromVariableId, fromEntity),
+                                    (EntityVariableOrFactReference<Solution_>) graph.lookup(toVariableId, toEntity));
+                        });
+                    });
+                    variableReferenceGraph.addAfterProcessor(sourceVariableId, (graph, toEntity) -> {
+                        alias.targetEntityFunctionStartingFromVariableEntity().accept(toEntity, fromEntity -> {
+                            graph.addEdge((EntityVariableOrFactReference<Solution_>) graph.lookup(fromVariableId, fromEntity),
+                                    (EntityVariableOrFactReference<Solution_>) graph.lookup(toVariableId, toEntity));
+                        });
+                    });
+                }
+                // Note: it is impossible to have a declarative variable affect graph edges,
+                // since accessing a declarative variable from another declarative variable is prohibited.
+            }
+        }
+
+        // Create the fixed edges in the graph
+        for (var entity : entities) {
+            for (var declarativeShadowVariableDescriptor : declarativeShadowVariableDescriptors) {
+                var entityClass = declarativeShadowVariableDescriptor.getEntityDescriptor().getEntityClass();
+                if (entityClass.isInstance(entity)) {
+                    var toVariableId = new VariableId(entityClass, declarativeShadowVariableDescriptor.getVariableName());
+                    for (var sourceRoot : declarativeShadowVariableDescriptor.getSources()) {
+                        for (var source : sourceRoot.variableSourceReferences()) {
+                            if (source.isTopLevel() && source.isDeclarative()) {
+                                var fromEntityClass = source.entityClass();
+                                var fromVariableName = source.variableName();
+                                var fromVariableId = new VariableId(fromEntityClass, fromVariableName);
+
+                                ((BiConsumer<Object, Consumer>) (BiConsumer) sourceRoot.valueEntityFunction())
```

Comment:
This is a strange cast. I don't think I've ever seen a double cast, honestly didn't know it was even possible. What's the point?